### DISCUSSION TITLE
Future Improvements 🚀

### DISCUSSION THREAD

**Discussion by tzanio** [GLVis/glvis#133] [Ideas]
What cool features will you like to see in GLVis in the future?

**Ideas:**

- [ ] Support for controlling line width, particularly in anti-aliased mode (options `-lw` and `-mslw`)
- [ ] Multi-touch support for phones and tablets
- [ ] Correct transparency, see https://github.com/GLVis/glvis/pull/166
- [ ] Improved performance: parallelize with OpenMP and/or MPI?
- [ ] Add a simple GUI, see https://github.com/wjakob/nanogui, https://github.com/ocornut/imgui
- [ ] Cutting plane control with 3D "gizmo" instead of keys (see [below](https://github.com/GLVis/glvis/discussions/133#discussioncomment-221224))
- [ ] Better shaders: tessellation, compute shaders, GPU acceleration, see https://github.com/jakubcerveny/hogtess
- [ ] Improve tessellation performance (see [below](https://github.com/GLVis/glvis/discussions/133#discussioncomment-221220))
- [ ] Support user-provided shaders
- [ ] Support Vulkan, Metal?
- [ ] Support multiple functions and operations on the same plot (see [below](https://github.com/GLVis/glvis/discussions/133#discussioncomment-356826))

**Done:**

- [x] Post Mac and Windows binaries on website
- [x] Release a signed version of the Mac app (see the [VisIt script](https://github.com/visit-dav/visit/blob/develop/src/tools/dev/masonry/masonry.py))
- [x] ~Live version of glvis.js on https://glvis.org/live~
- [x] ~Support for websockets, so the JS version can be used as drop-in replacement for the native GLVis server https://github.com/GLVis/glvis-js/pull/1#issuecomment-747564024~
- [x] ~Live demo of MFEM examples on its website, https://mfem.org/examples/~
- [x] ~Electron app, or other way for easy distribution~
- [x] ~Multi-threading support replacing fork()~

👇 Discuss

**Comment by tzanio** [comment]
cc: @kanye-quest, @tomstitt, @jakubcerveny, @v-dobrev, @jandrej

**Comment by jakubcerveny** [comment]
Improve tesselation performance (e.g. with OpenMP) and increase the tesselation limit. Right now larger solutions are very undertesselated by default. Also, the low default is enforced repeatedly when updating the solution.

**Comment by jakubcerveny** [comment]
Cutting plane keyboard control should be replaced/supplemented by a CAD-like 3D "gizmo" that would allow controlling the cutting plane with the mouse.

**Comment by tzanio** [comment]
Support multiple grid functions and visualization operations (e.g. cutting planes) on the same plot, see https://github.com/GLVis/glvis/issues/150.

One option could be a mode where GLVis can export the currently visible triangulation in a file plus an option to read and combine such triangulation files, cc: @v-dobrev

**Comment by tzanio** [comment]
Add support for controlling line width, particularly in anti-aliased mode (options `-lw` and `-mslw`)

**Comment by termi-official** [comment]
There is quite a big amount of code duplication between the `vssolution` and `vssolution3d` classes as well as between `vsvector` and `vsvector3d`. I think it is worth to refactor the duplicate code into base classes.

**Comment by CsatiZoltan** [comment]
Support VTK mixed meshes composed of elements with different polynomial degrees (e.g. Q2/T1, as in the attached mesh file).
[mesh.zip](https://github.com/GLVis/glvis/files/9746127/mesh.zip)

**Comment by termi-official** [comment]
What is the exact idea regarding the implementation of a simple GUI? Just replacing the colorbar and text with actual UI elements or do you have something more specific in mind?
